### ISSUE TITLE
OpenMM in FP16 or FP8 floating-point precision

### ISSUE THREAD

**Issue by razvanmarinescu** [openmm/openmm#4553]
Hi,

Has anyone tried OpenMM in floating-point precision lower than FP32? Can one still run simulations in FP16 or FP8? Which operations could be ideally moved to lower precision?

Thanks!

**Comment by peastman** [comment]
Nothing really, unfortunately.  There are a few very limited places where we make use of fp16 when building neighbor lists, but it isn't accurate enough for anything else.  Even working in fp32 requires a lot of care to maintain accuracy.  And people often use mixed precision, which uses fp64 in places where it makes the most difference.

**Comment by andychenbruce** [comment]
I am working on implementing f16. It seems the `src` in `CudaContext.cpp` passed to the compiler needs to have a `#include <cuda_fp16.h>` for half precision support, but it compiles the kernels at runtime with the `nvrtcCompileProgram` function and fails with 
```
Error compiling program: default_program(3): catastrophic error: could not open source file "cuda_fp16.h" (no directories in search list)
```
Can I switch the compilation of the CUDA kernels to compile-time rather than runtime? Or is there a reason why it is runtime compiled.

**Comment by peastman** [comment]
What specifically are you trying to implement?

**Comment by andychenbruce** [comment]
I am working on allowing a option for the simulations to use CUDA with half precision.

I just saw [this comment](https://github.com/openmm/openmm/blob/51a112a336db6cc0aea1c32b90d9d0a1f262cbfd/platforms/cuda/src/kernels/findInteractingBlocks.cu#L4-L9) in the CUDA kernel saying:

```
/**
 * To use half precision, we're supposed to include cuda_fp16.h.  Unfortunately,
 * it isn't included in the search path automatically, and there's no reliable
 * way to find where it's located on disk.  Instead we provide our own definitions
 * for the few symbols we need.
 */
 ```

I have found these [lines](https://github.com/openmm/openmm/blob/51a112a336db6cc0aea1c32b90d9d0a1f262cbfd/platforms/cuda/src/CudaContext.cpp#L473-L496) and others in the file that modify `src`, a string, which is later passed to `nvrtcCompileProgram` [here](https://github.com/openmm/openmm/blob/51a112a336db6cc0aea1c32b90d9d0a1f262cbfd/platforms/cuda/src/CudaContext.cpp#L571). I believe f16 support will be easiest if `half3` and `half4` vectors are written.

For F16 support I believe CUDA should use `half` documented [here](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__HALF.html) which I think needs `cuda_fp16.h` included in the C++ and CUDA code, but could be wrong as I have never written CUDA before. According to the comment it seems impossible to include the file at runtime. Otherwise I could just manually copying the `half` implementation into a `.cu` file similar to what that comment is describing above..

**Comment by peastman** [comment]
Like I noted above, there really is nothing for which fp16 is accurate enough.  Even getting away with fp32 takes a lot of care.

**Comment by andychenbruce** [comment]
Oh I see. You're saying its practically impossible for it to be useful?

**Comment by peastman** [comment]
Correct.  The lack of fp16 isn't because we haven't bothered to add it.  It's because we've investigated and concluded it wouldn't work.

**Comment by andychenbruce** [comment]
I think there is actually a use case for F16 support in generating training data for an AI to predict MD, and the lower accuracy may be stable enough to still train NN potentials. I will test if it will be too unstable even for this.

**Comment by peastman** [comment]
I think you'll find it's much too unstable, and I certainly wouldn't trust any results it produced.  With such low accuracy forces, it won't explore the correct energy landscape.

**Comment by andychenbruce** [comment]
I got some of the Cuda test to run but not pass with f16 on this [fork](https://github.com/andychenbruce/openmm-quantized). All other test seem to pass normally when I run `make test`, including Cuda and OpenCL tests, so the changes don't seem to have broken anything.

For example on my GPU when I run `./TestCudaCustomIntegrator f16` I get 
```
exception: Assertion failure at TestCustomIntegrator.h:98.  Expected 0.125, found 2.65846e+36
```

Another example is when I run `./TestCudaBrownianIntegrator f16`:
```
exception: Assertion failure at TestBrownianIntegrator.h:71.   Expected [-0.976209, 0, 0], found [-1, 0, nan]
```

And when I do `./TestCudaRandom f16`:
```
exception: Error downloading array force: CUDA_ERROR_INVALID_VALUE (1)
```
which makes me think I broke some of the memory layout or something.

I made the example `HelloArgon.cpp` run with Cuda [here](https://github.com/andychenbruce/openmm-quantized/blob/eddf7b32b6f6e395c15da5349ad8b21f4e546ba3/platforms/cuda/tests/TestCudaAndy.cpp) which runs normally without fp16, but instantly has one of the positions of the atoms become `nan`. I am currently trying to track it down in the Cuda code to see if there is some divide by zero due to inaccuracy, but so far I haven't found how it becomes `nan`. It may be some bug in the C++ code that I broke.

```
$ ./TestCudaAndy f16
REMARK  Using OpenMM platform CUDA
MODEL     1
ATOM      1  AR   AR     1       0.000   0.000   0.000  1.00  0.00
ATOM      2  AR   AR     1       5.000   0.000   0.000  1.00  0.00
ATOM      3  AR   AR     1      10.000   0.000   0.000  1.00  0.00
ENDMDL
MODEL     2
ATOM      1  AR   AR     1       0.000   0.000     nan  1.00  0.00
ATOM      2  AR   AR     1       5.000   0.000   0.000  1.00  0.00
ATOM      3  AR   AR     1      10.000   0.000   0.000  1.00  0.00
ENDMDL
MODEL     3
ATOM      1  AR   AR     1       0.000   0.000     nan  1.00  0.00
ATOM      2  AR   AR     1       5.000   0.000   0.000  1.00  0.00
ATOM      3  AR   AR     1      10.000   0.000   0.000  1.00  0.00
ENDMDL
...
...
...
``` 

I ended up  hard-coding the include path for `cuda_fp16.h` [here](https://github.com/andychenbruce/openmm-quantized/blob/eddf7b32b6f6e395c15da5349ad8b21f4e546ba3/platforms/cuda/src/CudaContext.cpp#L697).

I am also confused as to this part of the code [here](https://github.com/andychenbruce/openmm-quantized/blob/eddf7b32b6f6e395c15da5349ad8b21f4e546ba3/platforms/cuda/src/CudaContext.cpp#L475-L481) and [here](https://github.com/andychenbruce/openmm-quantized/blob/eddf7b32b6f6e395c15da5349ad8b21f4e546ba3/platforms/cuda/src/CudaContext.cpp#L495-L502). At the first link I just cargo cult-ed and put `8` for half after the previous `4` for double and `6` for single but I have no idea what it should be. At the second link I packed 2 half2's right next to each other in memory to emulate a half4.

**Comment by mattwthompson** [comment]
"Stability" as used above probably refers to the scientific results of using MD to look at dynamics, structure, etc. - if the forces applied accurately represent the derivative of the potentials, if the integrator and temperature and pressure control reproduce the physics, etc. I don't know how much precision needed for neural nets to pass benchmarks or for OpenMM to build, but for MD to be useful it has to respect some amount of baseline physics

**Comment by andychenbruce** [comment]
> I don't know how much precision needed for neural nets to pass benchmarks or for OpenMM to build, but for MD to be useful it has to respect some amount of baseline physics

The grad students in my group agree that fp16 is almost certainly not stable even for just 1 frame, but wanted to attempt it anyways as the speedup for training their neural net project would be very useful. If I understand correctly they only need one or a couple frames possibly at a very small Δt to be useful, as long as it is close to double/mixed/single precision. 

I can't confirm or deny if it meets these requirements yet as I haven't been able to figure out whether the `nan`s come from fp16 inaccuracies causing a divide by zero in Cuda which may be unavoidable, or if the C++ code has some bug like copying the buffer from Cuda incorrectly or something.
